### PR TITLE
EOS-20963 - Updated conf_cli.py for diff of all the keys between two files

### PR DESCRIPTION
**Pull request by balajiramachandran-seagate**
Signed-off-by: Balaji Ramachandran <balaji.ramachandran@seagate.com>

# Problem Statement
- Problem statement 
Unable to find the difference in the Keys between two files

# Design
-  For Bug describe the fix here. 
-  For Feature, Post the link to the solution page on the confluence CORTX Foundation Library 
- https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/284099843/CORTX+ISO+Package+Structure

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing 
- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: Y
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- [x] Confirm Testing was performed with installed RPM [Y/N]:  Y

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed 
- [x] Jira is updated 
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [x] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Comment by balajiramachandran-seagate** [comment]
Test cases:
Tested for build 552 and 61.
[root@ssc-vm-c-456 conf_store]# conf yaml:///tmp/RELEASE.INFO diff yaml:///tmp/RELEASE_1.INFO -k "BUILD"
1c1
< [552]
---
> [61]

Difference in the Keys between both the release.info files.
[root@ssc-vm-c-456 conf_store]# conf yaml:///tmp/RELEASE.INFO diff yaml:///tmp/RELEASE_1.INFO
1c1
< ['NAME', 'VERSION', 'BUILD', 'OS', 'DATETIME', 'KERNEL', 'COMPONENTS[0]', 'COMPONENTS[1]', 'COMPONENTS[2]', 'COMPONENTS[3]', 'COMPONENTS[4]', 'COMPONENTS[5]', 'COMPONENTS[6]', 'COMPONENTS[7]', 'COMPONENTS[8]', 'COMPONENTS[9]', 'COMPONENTS[10]', 'COMPONENTS[11]', 'COMPONENTS[12]', 'COMPONENTS[13]', 'COMPONENTS[14]']
---
> ['NAME', 'VERSION', 'BRANCH', 'THIRD_PARTY_VERSION', 'BUILD', 'OS', 'DATETIME', 'KERNEL', 'COMPONENTS[0]', 'COMPONENTS[1]', 'COMPONENTS[2]', 'COMPONENTS[3]', 'COMPONENTS[4]', 'COMPONENTS[5]', 'COMPONENTS[6]', 'COMPONENTS[7]', 'COMPONENTS[8]', 'COMPONENTS[9]', 'COMPONENTS[10]', 'COMPONENTS[11]', 'COMPONENTS[12]', 'COMPONENTS[13]', 'COMPONENTS[14]', 'COMPONENTS[15]', 'COMPONENTS[16]']

**Comment by balajiramachandran-seagate** [comment]
Test case Screen shot:

image

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 1]
```diff
@@ -229,3 +234,5 @@ def main():
 if __name__ == "__main__":
     rc = main()
     sys.exit(rc)
+
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:238]: Remove extra lines

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 2]
```diff
@@ -73,10 +73,15 @@ def get(args) -> str:
...
-            #todo the diff of all the keys between two files
-            sys.exit(0)
+            args.format = None
+            args.key_index = None
+            string_1 = ConfCli.get_keys(args)
+            ConfCli._index = "string_diff"
+            args.url = args.second_url
+            ConfCli.init(args.url)
+            string_2 = ConfCli.get_keys(args)
+            cmd = """bash -c "diff <(echo \\"%s\\") <(echo \\"%s\\")" """ %(string_1, string_2)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:84]: This line can be common for if and else.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 1 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py]: Removed extra lines

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 2 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py]: Updated the line common for both if and else.
cmd = """bash -c "diff <(echo \\"%s\\") <(echo \\"%s\\")" """ %(string_1, string_2)

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 3]
```diff
@@ -73,23 +73,27 @@ def get(args) -> str:
...
     def diff(args) -> str:
         """ Compare two diffenent string value for the given keys """
-        output = ""
         if len(args.args) < 1:
-            #todo the diff of all the keys between two files
-            sys.exit(0)
+            args.format = None
+            args.key_index = None
+            string_1 = ConfCli.get_keys(args)
+            ConfCli._index = "string_diff"
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:80]: You can not overwrite _index as it is initialized at the time of init which is one time activity.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 2 above)
Comment (venkuppu-chn) [reply py-utils/src/utils/conf_store/conf_cli.py]: Done. LGTM

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 3 above)
Comment (venkuppu-chn) [reply py-utils/src/utils/conf_store/conf_cli.py:80]: Yes correct if we are processing one file at a time. If we have to process two files we have to set the index twice.

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 1 above)
Comment (venkuppu-chn) [reply py-utils/src/utils/conf_store/conf_cli.py]: Done.

**Comment by balajiramachandran-seagate** [comment]
Output:

image

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 3 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py:80]: Done. LGTM

Comment (ujjwalpl) [reply py-utils/src/utils/conf_store/conf_cli.py:80]: I still see _index being modified. we can not modify _index. Its reserved for current config only.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Comment by balajiramachandran-seagate** [comment]
@ujjwalpl , Updated  _index . Please review once and let me know for any changes.

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 4]
```diff
@@ -128,13 +132,15 @@ def delete(args):
             Conf.save(ConfCli._index)
 
     @staticmethod
-    def get_keys(args) -> list:
+    def get_keys(args, index: str = None) -> list:
```
Comment (balajiramachandran-seagate) [inline py-utils/src/utils/conf_store/conf_cli.py:135]: Updated with index parameter

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 5]
```diff
@@ -128,13 +132,15 @@ def delete(args):
...
         """ Returns list of keys present in store """
         key_index = 'true' if args.key_index == None else args.key_index.lower().strip()
         key_index = True if key_index == 'true' else False if key_index == 'false' else None
-        if key_index == None:
+        if key_index is None:
```
Comment (balajiramachandran-seagate) [inline py-utils/src/utils/conf_store/conf_cli.py:139]: Updated the syntax.

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 6]
```diff
@@ -128,13 +132,15 @@ def delete(args):
...
             raise ConfError(errno.EINVAL, "invalid key_index value %s", key_index)
-        return Conf.get_keys(ConfCli._index, key_index=key_index)
+	if index is None:
+		index = ConfCli._index
+        return Conf.get_keys(index, key_index=key_index)
```
Comment (balajiramachandran-seagate) [inline py-utils/src/utils/conf_store/conf_cli.py:143]: Added the if condition

**Review by balajiramachandran-seagate — COMMENTED** [review]
@sachinpunadikar, Please review the changes.

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 4 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py:135]: @sachinpunadikar , Pls review the changes

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 5 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py:139]: @sachinpunadikar , Pls review the changes

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 6 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py:143]: @sachinpunadikar , Pls review the changes

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 7]
```diff
@@ -78,23 +78,27 @@ def get(args) -> str:
...
         if len(args.args) < 1:
-            #todo the diff of all the keys between two files
-            sys.exit(0)
+            args.format = None
+            args.key_index = None
+            string_1 = ConfCli.get_keys(args)
+            diff_index = "string_diff"
+            args.url = args.second_url
+            ConfCli.load(args.url, diff_index)
+            string_2 = ConfCli.get_keys(args)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_cli.py:88]: string_2 = ConfCli.get_keys(args, diff_index)

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 7 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py]: @ujjwalpl , Updated the changes

**Review by ujjwalpl — APPROVED** [review]

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/test/conf_store/test_conf_cli.py [hunk 8]
```diff
@@ -307,9 +317,7 @@ def test_conf_cli_merge(self):
     def tearDownClass(cls):
         delete_files()
 
-
```
Comment (sachinpunadikar) [inline py-utils/test/conf_store/test_conf_cli.py:310]: Keep this new line  .. there shd be 2 blank lines after class definition

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 9]
```diff
@@ -78,23 +78,27 @@ def get(args) -> str:
     @staticmethod
     def diff(args) -> str:
         """ Compare two diffenent string value for the given keys """
-        output = ""
         if len(args.args) < 1:
-            #todo the diff of all the keys between two files
-            sys.exit(0)
+            args.format = None
```
Comment (sachinpunadikar) [inline py-utils/src/utils/conf_store/conf_cli.py:82]: its repeat in if & else block. Take it out

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
Please work on the comments

**Comment by sachinpunadikar** [comment]
@rahul-tripathi-git please review this PR. It has reference to get_keys

File: py-utils/test/conf_store/test_conf_cli.py [hunk 10]
```diff
@@ -307,9 +317,7 @@ def test_conf_cli_merge(self):
...
 if __name__ == '__main__':
     # create the file and load sample json into it. Start test
     setup_and_generate_sample_files()
     unittest.main()
-
```
Comment (suryakumar1024) [inline py-utils/test/conf_store/test_conf_cli.py:315]: We should keep blank links at end of every script files.

File: py-utils/test/conf_store/test_conf_cli.py [hunk 11]
```diff
@@ -291,6 +291,16 @@ def test_conf_cli_properties_wrong_format_kv(self):
         except Exception:
             self.assertEqual(result_data[2], '22')
 
+    def test_conf_cli_by_get_diff_keys(self):
+        """
+        Test by retrieving a value using get api
+        """
+        cmd = "conf json:///tmp/file1.json diff json:///tmp/file2.json"
```
Comment (suryakumar1024) [inline py-utils/test/conf_store/test_conf_cli.py:298]: Kindly add test cases for negative scenarios

Comment (suryakumar1024) [reply py-utils/test/conf_store/test_conf_cli.py:298]: What if we passed like below
1) "conf json:///tmp/file1.json diff "
2) "conf json:///tmp/file1.json diff json:///tmp/not exist file"

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 12]
```diff
@@ -128,13 +132,15 @@ def delete(args):
             Conf.save(ConfCli._index)
 
     @staticmethod
-    def get_keys(args) -> list:
+    def get_keys(args, index: str = None) -> list:
         """ Returns list of keys present in store """
         key_index = 'true' if args.key_index == None else args.key_index.lower().strip()
```
Comment (suryakumar1024) [inline py-utils/src/utils/conf_store/conf_cli.py:137]: Shall we update this 'true' to True

File: py-utils/src/utils/conf_store/conf_cli.py [hunk 13]
```diff
@@ -128,13 +132,15 @@ def delete(args):
...
         key_index = True if key_index == 'true' else False if key_index == 'false' else None
```
Comment (suryakumar1024) [inline py-utils/src/utils/conf_store/conf_cli.py:138]: Same here we update this 'true' to True.

Kindly let me know from where do we get 'false' value here

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Comment by rahul-tripathi-git** [comment]
Hi @sachinpunadikar 
By default it will take the default deep_scan=True.

File: py-utils/test/conf_store/test_conf_cli.py (hunk 8 above)
Comment (balajiramachandran-seagate) [reply py-utils/test/conf_store/test_conf_cli.py:310]: Done

File: py-utils/src/utils/conf_store/conf_cli.py (hunk 9 above)
Comment (balajiramachandran-seagate) [reply py-utils/src/utils/conf_store/conf_cli.py:82]: Done

File: py-utils/test/conf_store/test_conf_cli.py (hunk 10 above)
Comment (balajiramachandran-seagate) [reply py-utils/test/conf_store/test_conf_cli.py:315]: Done

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by suryakumar1024 — APPROVED** [review]

**Review by s-arya — APPROVED** [review]

**Comment by mukul-seagate11** [comment]
@Gatekeepers, can this PR be merged as its already approved?
